File: src/openrtm_aist_python/OpenRTM_aist/Process.py
#!/usr/bin/env python
# -*- coding: euc-jp -*-


##
# \file Process.py
# \brief Process handling functions
# \author Noriaki Ando <n-ando@aist.go.jp> and Shinji Kurihara
#
# Copyright (C) 2010
#     Noriaki Ando
#     Task-intelligence Research Group,
#     Intelligent Systems Research Institute,
#     National Institute of
#         Advanced Industrial Science and Technology (AIST), Japan
#     All rights reserved.

import os,sys
import traceback
import subprocess

##
# @if jp
# @brief プロセスを起動する
# @else
# @brief Launching a process
# @endif
#
# int launch_shell(std::string command)
def launch_shell(command):
  args = command.split(" ")

  if sys.platform == "win32":
    CREATE_NEW_PROCESS_GROUP = 0x00000200
    subproc_args = { 'stdin':     None,
                     'stdout':    None,
                     'stderr':    None,
                     'cwd':       None,
                     'close_fds': False,
                     'creationflags': CREATE_NEW_PROCESS_GROUP}
  else:
    subproc_args = { 'stdin':     None,
                     'stdout':    None,
                     'stderr':    None,
                     'cwd':       None,
                     'close_fds': False,
                     'preexec_fn': os.setsid}

  try:
    p = subprocess.Popen(args, **subproc_args)
  except OSError:
    # fork failed
    if sys.version_info[0:3] >= (2, 4, 0):
      print traceback.format_exc()
    else:
      _exc_list = traceback.format_exception(*sys.exc_info())
      print "".join(_exc_list)

    return -1
  return 0

File: src/openrtm_aist/src/lib/coil/posix/coil/Process.cpp
// -*- C++ -*-
/*!
 * @file  Process.h
 * @brief Process handling functions
 * @date  $Date$
 * @author Noriaki Ando <n-ando@aist.go.jp>
 *
 * Copyright (C) 2010
 *     Noriaki Ando
 *     Task-intelligence Research Group,
 *     Intelligent Systems Research Institute,
 *     National Institute of
 *         Advanced Industrial Science and Technology (AIST), Japan
 *     All rights reserved.
 *
 * $Id$
 *
 */

#ifndef COIL_PROCESS_H
#define COIL_PROCESS_H

#include <stdlib.h>
#include <unistd.h>
#include <libgen.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <coil/stringutil.h>

namespace coil
{

  /*!
   * @if jp
   * @brief プロセスを起動する
   * @else
   * @brief Launching a process
   * @endif
   */
  int launch_shell(std::string command)
  {
    signal(SIGCHLD, SIG_IGN);

    pid_t pid;
    if((pid = fork()) < 0 )
      { // fork failed
        return -1; 
      }
    
    if (pid == 0) // I'm child process
      {
        //        signal(SIGCHLD, SIG_IGN);
        //        signal(SIGALRM, SIG_IGN);
        //        signal(SIGHUP , SIG_IGN);
        //        signal(SIGPIPE, SIG_IGN);
        //        signal(SIGTERM, SIG_IGN);
        setsid();
        //        close(0);
        //        close(1);
        //        close(2);
        //        open("/dev/null", O_RDWR);
        //        dup2(0, 1);
        //        dup2(0, 2);
        //        umask(0);

        coil::vstring vstr(::coil::split(command, " "));
        char* const * argv = ::coil::toArgv(vstr);

        execvp(vstr.front().c_str(), argv);
        
        return -1;
      }
    return 0;
  }

  int daemon(int nochdir, int noclose)
  {
    return daemon(nochdir, noclose);
  }


}; // namespace coil
#endif // COIL_PROCESS_H
